test/suites/integration/nodeclass_test.go

```diff
@@ -0,0 +1,108 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration_test
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+	"github.com/google/uuid"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	v1 "github.com/aws/karpenter-provider-aws/pkg/apis/v1"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("NodeClass IAM Permissions", func() {
+	var (
+		roleName   string
+		policyName string
+	)
+
+	DescribeTable("IAM Permission Failure Tests",
+		func(action string) {
+			policyDoc := fmt.Sprintf(`{
+            "Version": "2012-10-17",
+            "Statement": [
+                {
+                    "Effect": "Deny",
+                    "Action": ["%s"],
+                    "Resource": "*"
+                }
+            ]
+        }`, action)
+
+			deployment := &appsv1.Deployment{}
+			err := env.Client.Get(env.Context, types.NamespacedName{
+				Namespace: "kube-system",
+				Name:      "karpenter",
+			}, deployment)
+
+			Expect(err).To(BeNil())
+
+			sa := &corev1.ServiceAccount{}
+			err = env.Client.Get(env.Context, types.NamespacedName{
+				Namespace: "kube-system",
+				Name:      deployment.Spec.Template.Spec.ServiceAccountName,
+			}, sa)
+			Expect(err).To(BeNil())
+
+			roleName = strings.Split(sa.Annotations["eks.amazonaws.com/role-arn"], "/")[1]
+			policyName = fmt.Sprintf("TestPolicy-%s", uuid.New().String())
+
+			_, err = env.IAMAPI.PutRolePolicy(env.Context, &iam.PutRolePolicyInput{
+				RoleName:       aws.String(roleName),
+				PolicyName:     aws.String(policyName),
+				PolicyDocument: aws.String(policyDoc),
+			})
+			Expect(err).To(BeNil())
+
+			DeferCleanup(func() {
+				_, err := env.IAMAPI.DeleteRolePolicy(env.Context, &iam.DeleteRolePolicyInput{
+					RoleName:   aws.String(roleName),
+					PolicyName: aws.String(policyName),
+				})
+				Expect(err).To(BeNil())
+			})
+
+			env.ExpectCreated(nodeClass)
+			Eventually(func(g Gomega) {
+				env.ExpectUpdated(nodeClass)
+				g.Expect(nodeClass.StatusConditions().Get(v1.ConditionTypeValidationSucceeded).IsTrue()).To(BeFalse())
+
+				condition := nodeClass.StatusConditions().Get(v1.ConditionTypeValidationSucceeded)
+				g.Expect(condition).ToNot(BeNil())
+				g.Expect(condition.Message).To(ContainSubstring("unauthorized operation"))
```

Comment:
Might be nice to check the reason rather than the message (more consistent check since it's an actual error code)